### PR TITLE
Implement Exchange Step double signature validation

### PR DESCRIPTION
**Pull request by duliomatos1**
* Use new function `verifyExchanteStepSignatures()` from
consent-signaling-client to implement validation at
`ExchangeStepValidatorImpl` class

---
This change is [Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/121)

### CODE REVIEW

**Review by efoxepstein — CHANGES_REQUESTED** [review]
> Reviewed 9 of 9 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/121)* status: all files reviewed, 1 unresolved discussion (waiting on @duliomatos1, @jonmolle, and @stevenwarejones)

___
*[src/main/kotlin/org/wfanet/panelmatch/client/deploy/ExchangeWorkflowDaemon.kt, line 51 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/121#-MjjZocg8M83tKfDlAiy:-MjjZoch2aJ0YO4hvrqy:b5u4fq0) ([raw file](https://github.com/world-federation-of-advertisers/panel-exchange-client/blob/0e8dd9f51d6c7f63cb4939b086788ec5832ba74f/src/main/kotlin/org/wfanet/panelmatch/client/deploy/ExchangeWorkflowDaemon.kt#L51)):*
> ```Kotlin
>   abstract val privateStorage: VerifiedStorageClient
> 
>   abstract val dataProviderCertificate: X509Certificate
> ```

In general a single daemon will be handling exchanges with many partners (rather than just one EDP or one MP)

**Review by jonmolle — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/121)* status: all files reviewed, 2 unresolved discussions (waiting on @duliomatos1, @efoxepstein, and @stevenwarejones)

___
*[src/main/kotlin/org/wfanet/panelmatch/client/deploy/ExchangeWorkflowDaemon.kt, line 51 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/121#-MjjZocg8M83tKfDlAiy:-Mjjckq-036xIcqfqdZF:b-z4tfjr) ([raw file](https://github.com/world-federation-of-advertisers/panel-exchange-client/blob/0e8dd9f51d6c7f63cb4939b086788ec5832ba74f/src/main/kotlin/org/wfanet/panelmatch/client/deploy/ExchangeWorkflowDaemon.kt#L51)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

In general a single daemon will be handling exchanges with many partners (rather than just one EDP or one MP)
</blockquote></details>

In fact, I think right we take in a certificate bundle of trusted certs. I think this is fine for the moment, as I'm adding some unpacking there myself based on some not-yet-committed workflow wrapper fields.
___
*[src/main/kotlin/org/wfanet/panelmatch/client/launcher/ExchangeStepValidatorImpl.kt, line 25 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/121#-Mjjduug6OJJq-tL6SE7:-Mjjduug6OJJq-tL6SE8:bnadsvr) ([raw file](https://github.com/world-federation-of-advertisers/panel-exchange-client/blob/0e8dd9f51d6c7f63cb4939b086788ec5832ba74f/src/main/kotlin/org/wfanet/panelmatch/client/launcher/ExchangeStepValidatorImpl.kt#L25)):*
> ```Kotlin
> 
> class ExchangeStepValidatorImpl(
>   partyType: ExchangeWorkflow.Party,
> ```

Is there a need to know what party you are? I would think you should have both your and your opposite's cert, so you can just check them both.

**Review by efoxepstein — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/121)* status: all files reviewed, 1 unresolved discussion (waiting on @duliomatos1 and @stevenwarejones)

___
*[src/main/kotlin/org/wfanet/panelmatch/client/deploy/ExchangeWorkflowDaemon.kt, line 51 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/121#-MjjZocg8M83tKfDlAiy:-MjjpOZeDHQS2dubehz4:brfe6u2) ([raw file](https://github.com/world-federation-of-advertisers/panel-exchange-client/blob/0e8dd9f51d6c7f63cb4939b086788ec5832ba74f/src/main/kotlin/org/wfanet/panelmatch/client/deploy/ExchangeWorkflowDaemon.kt#L51)):*
<details><summary><i>Previously, jonmolle wrote…</i></summary><blockquote>

In fact, I think right we take in a certificate bundle of trusted certs. I think this is fine for the moment, as I'm adding some unpacking there myself based on some not-yet-committed workflow wrapper fields.
</blockquote></details>

OK, @duliomatos1, can you add a TODO somewhere around here to think about how multiple partners are supported?

**Review by duliomatos1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/121)* status: all files reviewed, 1 unresolved discussion (waiting on @stevenwarejones)

___
*[src/main/kotlin/org/wfanet/panelmatch/client/deploy/ExchangeWorkflowDaemon.kt, line 51 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/121#-MjjZocg8M83tKfDlAiy:-MjkRt_XE8BjBtK2WSWn:bvq8gav) ([raw file](https://github.com/world-federation-of-advertisers/panel-exchange-client/blob/0e8dd9f51d6c7f63cb4939b086788ec5832ba74f/src/main/kotlin/org/wfanet/panelmatch/client/deploy/ExchangeWorkflowDaemon.kt#L51)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

OK, @duliomatos1, can you add a TODO somewhere around here to think about how multiple partners are supported?
</blockquote></details>

Sure thing.

Right now, [SignedExchangeWorkflow](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/main/src/main/proto/wfa/measurement/api/v2alpha/exchange_step.proto#L44) does indicates just one EDP and one MP, could put a TODO in the proto as well as here
___
*[src/main/kotlin/org/wfanet/panelmatch/client/launcher/ExchangeStepValidatorImpl.kt, line 25 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/121#-Mjjduug6OJJq-tL6SE7:-MjkQfMP9rrisK3w2nq9:bt2p4zy) ([raw file](https://github.com/world-federation-of-advertisers/panel-exchange-client/blob/0e8dd9f51d6c7f63cb4939b086788ec5832ba74f/src/main/kotlin/org/wfanet/panelmatch/client/launcher/ExchangeStepValidatorImpl.kt#L25)):*
<details><summary><i>Previously, jonmolle wrote…</i></summary><blockquote>

Is there a need to know what party you are? I would think you should have both your and your opposite's cert, so you can just check them both.
</blockquote></details>

Right now we have a verification function under `MeasurementConsumerClient` and under `DataProviderClient`, but both are the same implementation in the end.

@stevenwarejones what would be your opinion here?

**Review by efoxepstein — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/121)* status: all files reviewed, 2 unresolved discussions (waiting on @duliomatos1, @jonmolle, and @stevenwarejones)

___
*[src/main/kotlin/org/wfanet/panelmatch/client/deploy/ExchangeWorkflowDaemon.kt, line 51 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/121#-MjjZocg8M83tKfDlAiy:-MjkZxir6bPbjeRxPria:b-2ifjqa) ([raw file](https://github.com/world-federation-of-advertisers/panel-exchange-client/blob/0e8dd9f51d6c7f63cb4939b086788ec5832ba74f/src/main/kotlin/org/wfanet/panelmatch/client/deploy/ExchangeWorkflowDaemon.kt#L51)):*
<details><summary><i>Previously, duliomatos1 (Dulio Matos) wrote…</i></summary><blockquote>

Sure thing.

Right now, [SignedExchangeWorkflow](https://github.com/world-federation-of-advertisers/cross-media-measurement-api/blob/main/src/main/proto/wfa/measurement/api/v2alpha/exchange_step.proto#L44) does indicates just one EDP and one MP, could put a TODO in the proto as well as here
</blockquote></details>

Yes, but this daemon will handle multiple SignedExchangeWorkflows

**Review by efoxepstein — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/121)* status: all files reviewed, 2 unresolved discussions (waiting on @duliomatos1, @jonmolle, and @stevenwarejones)

___
*[src/main/kotlin/org/wfanet/panelmatch/client/deploy/ExchangeWorkflowDaemon.kt, line 51 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/121#-MjjZocg8M83tKfDlAiy:-Mjk_5C26A_9rpqM0J3Y:b-v5ak74) ([raw file](https://github.com/world-federation-of-advertisers/panel-exchange-client/blob/0e8dd9f51d6c7f63cb4939b086788ec5832ba74f/src/main/kotlin/org/wfanet/panelmatch/client/deploy/ExchangeWorkflowDaemon.kt#L51)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

Yes, but this daemon will handle multiple SignedExchangeWorkflows
</blockquote></details>

In other words, as an MP, for example, you can run one daemon and handle exchanges with 150 different EDPs

**Review by stevenwarejones — CHANGES_REQUESTED** [review]
> Reviewed 8 of 9 files at r1.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/121)* status: all files reviewed, 2 unresolved discussions (waiting on @duliomatos1, @efoxepstein, and @jonmolle)

___
*[src/main/kotlin/org/wfanet/panelmatch/client/deploy/ExchangeWorkflowDaemon.kt, line 51 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/121#-MjjZocg8M83tKfDlAiy:-Mjkg5aa0cXJOAB0lQij:b-k2yqza) ([raw file](https://github.com/world-federation-of-advertisers/panel-exchange-client/blob/0e8dd9f51d6c7f63cb4939b086788ec5832ba74f/src/main/kotlin/org/wfanet/panelmatch/client/deploy/ExchangeWorkflowDaemon.kt#L51)):*
<details><summary><i>Previously, efoxepstein (Eli Fox-Epstein) wrote…</i></summary><blockquote>

In other words, as an MP, for example, you can run one daemon and handle exchanges with 150 different EDPs
</blockquote></details>

yeah, agree. add a todo. its a multi step process. you know your own certificate, of course. But, for the other party, you need to:
1. Look for work
2. Get the certificate for the other party for that work
3. Verify the workflow with that certificate
This applies to both MPs and EDPs.

Agree with Eli that you can just add it somewhere as a TODO 
___
*[src/main/kotlin/org/wfanet/panelmatch/client/launcher/ExchangeStepValidatorImpl.kt, line 25 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/121#-Mjjduug6OJJq-tL6SE7:-MjkfayJ6y3EBMojptr-:b-160f9p) ([raw file](https://github.com/world-federation-of-advertisers/panel-exchange-client/blob/0e8dd9f51d6c7f63cb4939b086788ec5832ba74f/src/main/kotlin/org/wfanet/panelmatch/client/launcher/ExchangeStepValidatorImpl.kt#L25)):*
<details><summary><i>Previously, duliomatos1 (Dulio Matos) wrote…</i></summary><blockquote>

Right now we have a verification function under `MeasurementConsumerClient` and under `DataProviderClient`, but both are the same implementation in the end.

@stevenwarejones what would be your opinion here?
</blockquote></details>

the way the consent signaling repo is currently set up, there are separate functions per party. Following that, there needs to be some kind of separate check for each party here.

**Review by jonmolle — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/121)* status: all files reviewed, 2 unresolved discussions (waiting on @duliomatos1, @efoxepstein, and @jonmolle)

___
*[src/main/kotlin/org/wfanet/panelmatch/client/launcher/ExchangeStepValidatorImpl.kt, line 25 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/121#-Mjjduug6OJJq-tL6SE7:-Mjl9W-iAvpRUqUAnvR-:bmxzq3r) ([raw file](https://github.com/world-federation-of-advertisers/panel-exchange-client/blob/0e8dd9f51d6c7f63cb4939b086788ec5832ba74f/src/main/kotlin/org/wfanet/panelmatch/client/launcher/ExchangeStepValidatorImpl.kt#L25)):*
<details><summary><i>Previously, stevenwarejones (Steven Ware Jones) wrote…</i></summary><blockquote>

the way the consent signaling repo is currently set up, there are separate functions per party. Following that, there needs to be some kind of separate check for each party here.
</blockquote></details>

Makes sense. LGTM then pending the other discussion.

**Review by stevenwarejones — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/121)* status: all files reviewed, 2 unresolved discussions (waiting on @duliomatos1, @efoxepstein, and @jonmolle)

**Review by duliomatos1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/121)* status: all files reviewed, 2 unresolved discussions (waiting on @duliomatos1, @efoxepstein, @jonmolle, and @stevenwarejones)

___
*[src/main/kotlin/org/wfanet/panelmatch/client/deploy/ExchangeWorkflowDaemon.kt, line 51 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/121#-MjjZocg8M83tKfDlAiy:-Mjow4vu43zkkI_XXARN:b-pwumqa) ([raw file](https://github.com/world-federation-of-advertisers/panel-exchange-client/blob/0e8dd9f51d6c7f63cb4939b086788ec5832ba74f/src/main/kotlin/org/wfanet/panelmatch/client/deploy/ExchangeWorkflowDaemon.kt#L51)):*
<details><summary><i>Previously, stevenwarejones (Steven Ware Jones) wrote…</i></summary><blockquote>

yeah, agree. add a todo. its a multi step process. you know your own certificate, of course. But, for the other party, you need to:
1. Look for work
2. Get the certificate for the other party for that work
3. Verify the workflow with that certificate
This applies to both MPs and EDPs.

Agree with Eli that you can just add it somewhere as a TODO 
</blockquote></details>

thanks for the context, will do

**Review by duliomatos1 — COMMENTED** [review]
> Dismissed @jonmolle from a discussion.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/121)* status: all files reviewed, 1 unresolved discussion (waiting on @efoxepstein, @jonmolle, and @stevenwarejones)

___
*[src/main/kotlin/org/wfanet/panelmatch/client/launcher/ExchangeStepValidatorImpl.kt, line 25 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/121#-Mjjduug6OJJq-tL6SE7:-MjowQFH3U0Hpjl0WwTO:b34v6zc) ([raw file](https://github.com/world-federation-of-advertisers/panel-exchange-client/blob/0e8dd9f51d6c7f63cb4939b086788ec5832ba74f/src/main/kotlin/org/wfanet/panelmatch/client/launcher/ExchangeStepValidatorImpl.kt#L25)):*
<details><summary><i>Previously, jonmolle wrote…</i></summary><blockquote>

Makes sense. LGTM then pending the other discussion.
</blockquote></details>

ok, will keep as is, and we can revisit with the TODO

**Review by duliomatos1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/121)* status: 7 of 9 files reviewed, all discussions resolved (waiting on @efoxepstein and @stevenwarejones)

___
*[src/main/kotlin/org/wfanet/panelmatch/client/deploy/ExchangeWorkflowDaemon.kt, line 51 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/121#-MjjZocg8M83tKfDlAiy:-MjpH2Dd39t5DMBdzd6w:b-6e05iz) ([raw file](https://github.com/world-federation-of-advertisers/panel-exchange-client/blob/0e8dd9f51d6c7f63cb4939b086788ec5832ba74f/src/main/kotlin/org/wfanet/panelmatch/client/deploy/ExchangeWorkflowDaemon.kt#L51)):*
<details><summary><i>Previously, duliomatos1 (Dulio Matos) wrote…</i></summary><blockquote>

thanks for the context, will do
</blockquote></details>

Done. (TODO is inside `ExchangeStepValidatorImpl`)

**Review by stevenwarejones — APPROVED** [review]
> Reviewed 1 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/panel-exchange-client/121)* status: 8 of 9 files reviewed, all discussions resolved (waiting on @efoxepstein and @stevenwarejones)
